sstables/partition.cc

```diff
@@ -433,7 +437,15 @@ class sstable_mutation_reader : public mp_row_consumer_reader {
                     if (is_buffer_full() || _partition_finished || _end_of_stream) {
                         return make_ready_future<>();
                     }
-                    return advance_context(_consumer.maybe_skip()).then([this] {
```

Comment:
I don't understand why you need this patch. The old code called `maybe_skip()`, which should tell advance_context() to do nothing if skipping wasn't really needed. Maybe the bug was in maybe_skip?

Reply:
Possible, I'll check. My understanding of the code is quite weak.

Reply:
advance_context calls index_reader::advance_to unconditionally:

```c++
    future<> advance_context(std::optional<position_in_partition_view> pos) {
        if (!pos || pos->is_before_all_fragments(*_schema)) {
            return make_ready_future<>();
        }
        assert (_current_partition_key);
        return [this] {
            if (!_index_in_current_partition) {
                _index_in_current_partition = true;
                return get_index_reader().advance_to(*_current_partition_key);
            }
            return make_ready_future();
        }().then([this, pos] {
            return get_index_reader().advance_to(*pos).then([this] {
                index_reader& idx = *_index_reader;
                auto index_position = idx.data_file_positions();
                if (!_context->need_skip(index_position.start)) {
                    return make_ready_future<>();
                }
                return _context->skip_to(idx.element_kind(), index_position.start).then([this, &idx] {
                    _sst->get_stats().on_partition_seek();
                    set_range_tombstone_start_from_end_open_marker(_consumer, *_schema, idx);
                });
            });
        });
    }
```

The first call to advance_to() is conditional, but the second is not.

Reply:
Actually, there's an early return up front. So maybe_skip() should have been enough.

Reply:
> I don't understand why you need this patch. The old code called `maybe_skip()`, which should tell advance_context() to do nothing if skipping wasn't really needed. Maybe the bug was in maybe_skip?

You were right, the regressions in perf_fast_forward disappear with patch 1 removed. But I get a failure in sstable_mutation_tests's test_no_index_reads_when_rows_fall_into_range_boundaries. Maybe it's a false positive?

Reply:
(the regression only appears in debug mode, presumably due to higher chances of preemption)

Reply:
Also appears in release mode with task quota set to 5us. But, it looks to me like a true positive, which I can't explain if maybe_skip() is correct.

Reply:
If you share the test log with the failure maybe I can guess something.

Reply:
```c++
            auto before = index_accesses();

            {
                assert_that(ms.make_reader(s, tests::make_permit()))
                    .produces(m1)
                    .produces(m2)
                    .produces_end_of_stream();

                BOOST_REQUIRE_EQUAL(index_accesses(), before);
            }
```

index_accesses() = 2, before == 0

Later I'll generate a trace-level log, see if I find anything, and post it here.